compiler/rustc_ast_lowering/src/format.rs

```diff
@@ -0,0 +1,354 @@
+    let format_placeholder_new = ctx.arena.alloc(ctx.expr_lang_item_type_relative(
+        sp,
+        hir::LangItem::FormatPlaceholder,
+        sym::new,
+    ));
+    let args = ctx.arena.alloc_from_iter([position, fill, align, flags, prec, width]);
+    ctx.expr_call_mut(sp, format_placeholder_new, args)
+}
+
+fn expand_format_args<'hir>(
+    ctx: &mut LoweringContext<'_, 'hir>,
+    macsp: Span,
+    fmt: &FormatArgs,
+) -> hir::ExprKind<'hir> {
+    let lit_pieces =
+        ctx.arena.alloc_from_iter(fmt.template.iter().enumerate().filter_map(|(i, piece)| {
+            match piece {
+                &FormatArgsPiece::Literal(s) => Some(ctx.expr_str(fmt.span, s)),
+                &FormatArgsPiece::Placeholder(_) => {
+                    // Inject empty string before placeholders when not already preceded by a literal piece.
+                    if i == 0 || matches!(fmt.template[i - 1], FormatArgsPiece::Placeholder(_)) {
+                        Some(ctx.expr_str(fmt.span, kw::Empty))
+                    } else {
+                        None
+                    }
+                }
+            }
+        }));
+    let lit_pieces = ctx.expr_array_ref(fmt.span, lit_pieces);
+
+    // Whether we'll use the `Arguments::new_v1_formatted` form (true),
+    // or the `Arguments::new_v1` form (false).
+    let mut use_format_options = false;
+
+    // Create a list of all _unique_ (argument, format trait) combinations.
+    // E.g. "{0} {0:x} {0} {1}" -> [(0, Display), (0, LowerHex), (1, Display)]
+    let mut argmap = FxIndexSet::default();
+    for piece in &fmt.template {
+        let FormatArgsPiece::Placeholder(placeholder) = piece else { continue };
+        if placeholder.format_options != Default::default() {
+            // Can't use basic form if there's any formatting options.
+            use_format_options = true;
+        }
+        if let Ok(index) = placeholder.argument.index {
+            if !argmap.insert((index, ArgumentType::Format(placeholder.format_trait))) {
+                // Duplicate (argument, format trait) combination,
+                // which we'll only put once in the args array.
+                use_format_options = true;
+            }
+        }
+    }
+
+    let format_options = use_format_options.then(|| {
+        // Generate:
+        //     &[format_spec_0, format_spec_1, format_spec_2]
+        let elements: Vec<_> = fmt
+            .template
+            .iter()
+            .filter_map(|piece| {
+                let FormatArgsPiece::Placeholder(placeholder) = piece else { return None };
+                Some(make_format_spec(ctx, macsp, placeholder, &mut argmap))
+            })
+            .collect();
+        ctx.expr_array_ref(macsp, ctx.arena.alloc_from_iter(elements))
+    });
+
+    let arguments = fmt.arguments.all_args();
+
+    // If the args array contains exactly all the original arguments once,
+    // in order, we can use a simple array instead of a `match` construction.
```

Comment:
Please document why we do this optimization at all, instead of always using the `match` version

Reply:
Hm, it might just have been to keep the `-Zunpretty=expanded` output less verbose. I'll go dig a bit through the git and issue history to find out.

Reply:
Yes, this is the original PR: https://github.com/rust-lang/rust/pull/91359

I think we care less about a verbose `-Zunpretty=hir` than `-Zunpretty=expanded`, so I'll just remove this.

Reply:
Oh actually, that PR did result in a lot of performance improvements: https://perf.rust-lang.org/compare.html?start=523be2e05da322daaecf1ecc8f2c0d625f5f46e3&end=0bcacb391b28460f5a50fd627f01f670dfcfc7cc

So let's run another perf test with this case removed, to see if it makes the same difference here at this level.

Reply:
Here are the perf results, comparing this PR with the 'simple array' special case, to a version of this PR with that special case removed: https://perf.rust-lang.org/compare.html?start=e36affffe97378a0027b4bcfbb18d27356164ed0&end=5dbee4d3a6728eb4530fb66c9775834438ecec74&stat=instructions:u

Unfortunately, all red numbers. :(

Reply:
Added a comment about this with a link to these perf results.